lib: os: cbprintf: Add new options to cbprintf packaging

nordic-krch opened the pull request:
Extended packaging by adding an option to create a package that does not contain transient strings copied into the package body. When flag `CBPRINTF_PACKAGE_ADD_RW_STR_POS` is passed to packaging (static or runtime), an array of transient string locations is added to the package instead of copying string body. That creates transient package which is valid as long as transient strings are valid but such package can be resolved to self-contained package.

After this change we can identify 3 variants of packages:

* Self-contained (default) - non read-only strings appended to the package. String can be formatted from such package as long as there is access to read-only string locations. Package may contain information where read-only strings are located within the package. That information can be used to convert packet to fully self-contained package.
* Fully self-contained - all strings are appended to the package. String can be formatted from such package without any external data.
* Transient- only arguments are stored. Package contain information where pointers to non read-only strings are located within the package. Optionally, it may contain read-only string location information. String can be formatted from such package as long as non read-only strings are still valid and read-only strings are accessible. Alternatively, package can be converted to self-contained package or fully self-contained if information about read-only string locations is present in the package.

Added function for copying packages (`cbprintf_package_copy()`). It accepts options to append read-write or read-only strings so it can transform more transient package to more self-contain version. Previously added `cbprintf_fsc_package()` is a copying with option to append read-only strings (and read-write strings). When output package is not provided to copying function it just calculates the length of the output package.

Most important, transient package can be created with `CBPRINTF_STATIC_PACKAGE()` (at compile time) so actual format string is not needed in the binary (fyi @dcpleung). It will significantly speed up logging when message contains string pointer arguments.

The usage for logging will be following:
- statically create transient package
- pass transient function to logging function which will first determine self-contain package size, allocate space for it, resolve transient package to self-contain package.

Copied https://github.com/zephyrproject-rtos/zephyr/commit/9d277468ac5f5a81d57440c2762c0bb1ec06443d to use it in the test.

nordic-krch committed:
lib: os: cbprintf: Extend API to support new packaging modes

Added new flags to packaging API:
- CBPRINTF_PACKAGE_ADD_RO_STR_IDXS - when set, read-only string
  locations are appended to the package
- CBPRINTF_PACKAGE_ADD_RW_STR_IDXS - when set, read-write string
  locations are appended to the package (instead of appending actual
  strings)
- CBPRINTF_PACKAGE_FIRST_RO_STR_CNT(n) - indicate that n first strings
  are known to be read only. Ignored in runtime packaging.

Add function for copying packages with optional appending strings.

Changed CBPRINTF_MUST_RUNTIME_PACKAGE to use same flags as packaging.

Aligned logging and test to those changes.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: reference: misc: Update cbprintf documentation

Added information about package variants and updated package
structure table.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

dcpleung commented:
(I have just skimmed through the code so my understanding may be off...)

* By default, static packaging is used which only store pointers to RW strings. With deferred mode, content of those strings may have changed between calling LOG_*() and actually being printed. Is this the case, or did I miss something?

* Dictionary logging, currently, simply dumps the whole package and let the offline python script deal with decoding the packages. So these changes may affect that. Though, as long as the RW strings are inside the packages, it should be fine.

* FYI, The MIPI Sys-T backend will need to support catalog messages (which is basically dictionary logging). However, their scheme and output require tagging each arguments. So basically, instead of `printk("%d %s", 1234, "test");`, you do `mipi_printk(<pointer to "%d %s", ARG_TYPE_INT, 1234, ARG_TYPE_C_STRING, "test", ARG_END);`. And they output the tags with each argument. I think this can be done via forcing runtime packaging with tagging similar to https://github.com/zephyrproject-rtos/zephyr/pull/42946, with corresponding changes in cbprintf to package those tags (and ignoring them when printing through cbvprintf). The MIPI backend can take the tagged packages and send them to the MIPI library. So there will probably be some invasive changes coming.

nordic-krch committed:
tests: unit: cbprintf: Align test to changes in cbprintf

Aligned tests to handle the fact that static packaging return
error when detects read-write string and storing read-write
positions is not enabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: cbprintf_package: Add test for CBPRINTF_MUST_RUNTIME_PACKAGE

Add test for CBPRINTF_MUST_RUNTIME_PACKAGE macro.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@dcpleung 
>By default, static packaging is used which only store pointers to RW strings. With deferred mode, content of those strings may have changed between calling LOG_*() and actually being printed. Is this the case, or did I miss something?

The idea is that static packaging will store RW pointers and locations. Then `z_log_msg2_finalize` which will do copying with RW resolving. So after applying this extension package that will end up in the logging buffer will be the same as now.

dcpleung commented:
Good to know. Thanks.

dcpleung commented on doc/reference/misc/formatted_output.rst:
```diff
@@ -113,9 +131,13 @@ formatting since address changes whenever package is copied.
 |                  +-------------------------------------------------------------------------+
 |                  | ...                                                                     |
 +------------------+-------------------------------------------------------------------------+
+| String location  | Indexes of words within the package were read-only strings are located  |
+| information      +-------------------------------------------------------------------------+
+| (optional)       | Indexes of words within the package were transient strings are located  |
```
s/were/where/?

nordic-krch commented:
I've created draft PR #43520 where logging is adapted. Using static packaging with resolving shows significant improvements (twice faster)  when logging messages with `%s`.

dcpleung reviewed (commented):
This is going to affect the existing dictionary logging as the current scheme is to dump the whole package, and process it offline.

nordic-krch commented on lib/os/cbprintf_packaged.c:
```diff
@@ -634,64 +715,196 @@ int cbpprintf_external(cbprintf_cb out,
-		out[1] = s_nbr + ros_nbr;
-		out[2] = 0;
-		out += args_size;
+	/* If flags are not set or appending requenst without without rw string
```
done

nordic-krch commented:
@dcpleung 
> This is going to affect the existing dictionary logging as the current scheme is to dump the whole package, and process it offline.

It shouldn't. Package that contains indexes to read write strings is transient (created on stack only) and is converted to "self-contained" version before it is committed to logging buffer, see https://github.com/zephyrproject-rtos/zephyr/blob/d0f14ec7bc71762baba74b9a154cdec1298c0eac/subsys/logging/log_msg2.c#L44

dcpleung commented:
I have tried dictionary logging and it still works with your changes. It's the documentation changes about the new optional bytes in the package that caused my confusion.

nordic-krch commented:
@npitre i was applying comments in documentation. Can you re-review.

I also added minor optimization to copying function to store calculated string lengths in the package size calculation phase and reuse it during copying. Test shows ~5% speedup in logging the message (for arbitrary string used in the test).

nordic-krch committed:
tests: lib: cbprintf_packaged: Extend test to check new modes

Extended test to check new flags added to cbprintf packaging API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@dcpleung  @npitre Had to rebase and resolve conflicts.
